### ISSUE TITLE
Buffer bug in filter function

### ISSUE THREAD

**Issue by flindvall** [phpseclib/mcrypt_compat#24]
When size of bucket data plus buffer divided by block length does not leave a remainder and buffer is not empty it needs to be cleared to avoid buffer content being reinserted next time filter function is called. Please consider this fix:
if ($extra) {
    $this->buffer = substr($bucket->data, -$extra);
    $bucket->data = substr($bucket->data, 0, -$extra);
} else if (strlen($this->buffer)) {
     $this->buffer = "";
}

**Comment by flindvall** [comment]
This is a cleaner fix:
if ($this->block_mode) {
    $bucket->data = $this->buffer . $bucket->data;
    $this->buffer = '';

**Comment by terrafrost** [comment]
So one thing that'd help is some code that'd enable me to reproduce the problem and verify the fix. In lieu of that I can attempt to reproduce the issue based on your description of the problem, but it'd be faster if you could provide that yourself lol.

Being able to reproduce the issue also enables the creation of a unit test.

**Comment by flindvall** [comment]
Yes, I agree that the proper procedure would be to provide code to reproduce the issue, however this is a logical flaw that can be described in words. I have encountered this when trying to decrypt data. The scenario was that 13 bytes were in the buffer and 611 bytes in $bucket->data. That makes 624 bytes total that is evenly divided by a block length of 16 and since $extra = strlen($bucket->data) % $this->block_length will be 0 then nothing will be done to take care of the 13 bytes in the buffer since the buffer is only adjusted if there is a remainder. The result is that the 13 bytes remains in the buffer and will cause the decryption to break on the next call to the filter function. Clearing the buffer after its content has been added to $bucket->data solves the problem. How about this test case with a file of size 1888 bytes:
$opts = array('iv'=>$iv, 'key'=>$key, 'mode'=>'cbc');
$fd = fopen($encrypted_file, 'rb');
stream_filter_append($fd, "mdecrypt.rijndael-128", STREAM_FILTER_READ, $opts);
$content .= fread($fd, 653);
$content .= fread($fd, 611);
$content .= fread($fd, 624);

**Comment by terrafrost** [comment]
Does https://github.com/phpseclib/mcrypt_compat/commit/0220d5916d5027b85370211e03231e1144c245c3 fix the issue for you?

It is quite difficult to unit test this as PHP doesn't really let you control (in-so-far as I know) how much of the file PHP actually reads. It doesn't look like it looks at the parameters that are passed to `fread`. Like 653 + 611 + 624 = 1888. So I tried to read the first 653 bytes and `$bucket->data` was already 1888 so I was unable to test the issue. This was true with `php://memory`, `php://temp` and an actual file.

I tried to have two file pointers - one that was write only and one that was read only. I then tried to write 653 bytes with the write-only file pointer and then read the 653 bytes with the read-only file pointer but locking interferes with that.
